### PR TITLE
feat(submodules): Initial support for git submodules

### LINKED ISSUE #2502: git submodule support
**What would you like Renovate to be able to do?**
Low priority, but just had the first use case ever for git submodules in one of my open source projects, would have loved to use Renovate.
https://github.com/felixfbecker/PowerShellXSD

**Describe the solution you'd like**
PRs to update git submodules

**Describe alternatives you've considered**
Currently using Dependabot

### PR DESCRIPTION
**Pull request by JamieMagee**
See JamieMagee/PowerShellXSD#2 for an example

**TODO**
- [ ] Clarify versioning
- [x] Add tests
- [x] Get test coverage to 100%
- [x] Test on a wider range of repositories
  - [PowerShellXSD](https://github.com/JamieMagee/PowerShellXSD/pull/2)
  - [lxqt](https://github.com/JamieMagee/lxqt/pulls) (Many submodules + relative URLs)
- [x] Caching

Fixes #2502

### CODE REVIEW

**Comment by JamieMagee** [comment]
I think I'm ready to open this up for review now. Still have some further testing to do, especially on platforms other than GitHub, but it appears to be working well.

File: lib/datasource/git-submodules/index.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import Git from 'simple-git/promise';
+import { URL } from 'url';
+
+import { ReleaseResult, PkgReleaseConfig, DigestConfig } from '../common';
+import { logger } from '../../logger';
+
+export async function getPkgReleases({
+  lookupName,
```
Comment (ViceIce) [inline lib/datasource/git-submodules/index.ts]:
```suggestion
  lookupName: string,
```

File: lib/datasource/git-submodules/index.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  registryUrls,
```
Comment (ViceIce) [inline lib/datasource/git-submodules/index.ts]:
```suggestion
  registryUrls: string[],
```

File: lib/manager/git-submodules/extract.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        currentValue,
+        currentDigest: currentValue,
+      };
+    })
+  );
+
+  return { deps, datasource: 'gitSubmodules' };
+}
+
+const getUrl = async (
```
Comment (ViceIce) [inline lib/manager/git-submodules/extract.ts]:
```suggestion
async function getUrl (
```

File: lib/manager/git-submodules/extract.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  git: Git.SimpleGit,
+  gitModulesPath: string,
+  submoduleName: string
+) => {
```
Comment (ViceIce) [inline lib/manager/git-submodules/extract.ts]:
```suggestion
): Promise<string> {
```

File: lib/manager/git-submodules/extract.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  }
+  const remoteUrl = (await git.raw([
+    'config',
+    '--get',
+    'remote.origin.url',
+  ])).trim();
+  return URL.resolve(`${remoteUrl}/`, path);
+};
+
+const getBranch = async (gitModulesPath: string, submoduleName: string) =>
```
Comment (ViceIce) [inline lib/manager/git-submodules/extract.ts]: Convert to function.

File: lib/platform/git/storage.ts [hunk 6]
```diff
@@ -125,6 +125,7 @@ export class Storage {
     const submodules = await this.getSubmodules();
     for (const submodule of submodules) {
       try {
+        logger.info(`Cloning git submodule at ${submodule}`);
```
Comment (ViceIce) [inline lib/platform/git/storage.ts]: Do we need this? Can we lower the level?

**Review by viceice — CHANGES_REQUESTED** [review]
Just type checking

File: lib/platform/git/storage.ts (hunk 6 above)
Comment (rarkins) [reply lib/platform/git/storage.ts]: Debug would be nice

File: lib/datasource/git-submodules/index.ts (hunk 1 above)
Comment (JamieMagee) [reply lib/datasource/git-submodules/index.ts]: Types come from `PkgReleaseConfig`. This is just destructing. Same on the line below.

Comment (ViceIce) [reply lib/datasource/git-submodules/index.ts]: Ups, you are right. Sorry.

File: lib/manager/git-submodules/artifacts.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { UpdateArtifactsConfig, UpdateArtifactsResult } from '../common';
+
+export default function updateArtifacts(
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply lib/manager/git-submodules/artifacts.ts]: The artifact is the directory that the `.gitmodules` file points to. For example, with the `.gitmodules` content

```
[submodule "abc"]
	path = abc
	url = git@github.com:abc/abc
```

Then we need to commit the directory `abc` to the branch. We only add package files or artifacts to a branch, so if `.gitmodules` is the package file, then the directory it points to feels like it should be the artifact.

File: lib/versioning/git/index.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import * as generic from '../loose/generic';
```
Comment (JamieMagee) [inline lib/versioning/git/index.ts]: I wanted to find opinions on the versioning scheme for git submodules (I don't think we have any other manager that only supports hashes).

As git submodules only supports hashes, I don't think any of the concepts around versioning really apply. You can't compare them, there's no compatability, etc. My thoughts were basically, if they're different, then we can update.

Comment (rarkins) [reply lib/versioning/git/index.ts]: Assuming it's logically possible, I would think that we don't need a concept of "versions" for git submodules and instead treat everything as a "digest update". Does that work?

File: lib/manager/git-submodules/artifacts.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/manager/git-submodules/artifacts.ts]: OK. I wasn't aware of the need to also add the directory too. I'm still a little confused by this git concept. I assume the `path` is constant and not something we change, but then I wonder why it needs updating unless git has duplicate references for submodules (once in `.gitmodules` and again within the directory somehow?)

[the remark this answers is not in the document]
Comment (JamieMagee) [reply lib/manager/git-submodules/artifacts.ts]: I think git is unique, in that we don't need to edit the package file (`.gitmodules`), but we do need to update the artifacts (The path from `.gitmodules`)

File: lib/versioning/git/index.ts (hunk 8 above)
Comment (JamieMagee) [reply lib/versioning/git/index.ts]: Again, I think git submodules are unique, in that they only have digests, and not explicit versions. Unfortunately all of our logic in Renovate is setup around supporting a version first, and a digest as something extra on top of the version.

This is why I've had to arbitrarily create a version for git submodules, even though it's actually a hash/digest.

**Review by viceice — APPROVED** [review]

**Comment by rarkins** [comment]
@JamieMagee something I meant to ask before I merged - what is the branch format/name? I hope it won't close/open/close/open PRs each time the submodule is updated?

**Comment by JamieMagee** [comment]
> @JamieMagee something I meant to ask before I merged - what is the branch format/name? I hope it won't close/open/close/open PRs each time the submodule is updated?

@rarkins Ah, the format is `<submodule-path>-digest`, so it does have the possibility to be quite noisy

**Pull request merged**